syncer/syncer.go

```diff
@@ -0,0 +1,402 @@
+// Configuration is the config params for syncer
+type Configuration struct {
+	SyncInterval time.Duration
+	// the sync process will try validate the current layer if ValidationDelta has elapsed.
+	ValidationDelta time.Duration
+	AlwaysListen    bool
+}
+
+const (
+	outOfSyncThreshold  types.LayerID = 2 // see notSynced
+	numGossipSyncLayers types.LayerID = 2 // see gossipSync
+)
+
+type syncState uint32
+
+const (
+	// notSynced is the state where the node is outOfSyncThreshold layers or more behind the current layer.
+	notSynced syncState = iota
+	// gossipSync is the state in which a node listens to at least one full layer of gossip before participating
+	// in the protocol. this is to protect the node from participating in the consensus without full information.
+	// for example, when a node wakes up in the middle of layer N, since it didn't receive all relevant messages and
+	// blocks of layer N, it shouldn't vote or produce blocks in layer N+1. it instead listens to gossip for all of
+	// layer N+1 and starts producing blocks and participates in hare committee in layer N+2
+	gossipSync
+	// synced is the phase where we are 2 layers or more behind the current layer
+	synced
+)
+
+func (s *syncState) String() string {
+	switch *s {
+	case 0:
+		return "notSynced"
+	case 1:
+		return "gossipSync"
+	case 2:
+		return "synced"
+	default:
+		return "unknown"
+	}
+}
+
+// Syncer is responsible to keep the node in sync with the network.
+type Syncer struct {
+	logger log.Log
+
+	conf      Configuration
+	ticker    layerTicker
+	mesh      *mesh.Mesh
+	fetcher   layerFetcher
+	syncOnce  sync.Once
+	syncState syncState
+	isBusy    uint32
+	syncTimer *time.Ticker
+	// targetSyncedLayer is used to signal at which layer we can set this node to synced state
+	targetSyncedLayer types.LayerID
+	// awaitSyncedCh is used to notify subscribes that this node is synced
+	awaitSyncedCh chan struct{}
+	awaitSyncedMu sync.Mutex
+
+	shutdownCtx context.Context
+	cancelFunc  context.CancelFunc
+
+	// recording the run # since started. for logging/debugging only.
+	run uint64
+}
+
+// NewSyncer creates a new Syncer instance.
+func NewSyncer(ctx context.Context, conf Configuration, ticker layerTicker, mesh *mesh.Mesh, fetcher layerFetcher, logger log.Log) *Syncer {
+	shutdownCtx, cancel := context.WithCancel(ctx)
+	return &Syncer{
+		logger:        logger,
+		conf:          conf,
+		ticker:        ticker,
+		mesh:          mesh,
+		fetcher:       fetcher,
+		syncState:     notSynced,
+		syncTimer:     time.NewTicker(conf.SyncInterval),
+		awaitSyncedCh: make(chan struct{}),
+		shutdownCtx:   shutdownCtx,
+		cancelFunc:    cancel,
+	}
+}
+
+// Close stops the syncing process and the goroutines syncer spawns.
+func (s *Syncer) Close() {
+	// TODO: ensure goroutines are all terminated before shutting down
+	s.cancelFunc()
+}
+
+// AwaitSynced returns the channel that will be notified whenever the node enters synced state.
+func (s *Syncer) AwaitSynced() chan struct{} {
```

Comment:
I think we should change this function to read from the channel and listen to result, since `s.awaitSyncedCh` can be switched out

Reply:
done. but changed to `RegisterChForSynced` and uses client's channel to notify. the only caller of this is activation.go and it really wants a channel to select on. and i don't want to change logic in activation.go either.